_blog/2023-09-13-building-web-scraper-in-rust.mdx

```diff
@@ -0,0 +1,381 @@
+    let productname_selector = Selector::parse("h2 > a").unwrap();
+    
+    let name = entry.select(&productname_selector).next().expect("Couldn't find the product name").text.next().unwrap().to_string();
+
+// Amazon products can have two prices: a current price, and an "old price". We iterate through both of these and map them to a Vec<String>.
+    let price_text = entry.select(&price_selector).map(|x| x.text().next().unwrap().to_string()).collect::<Vec<String>>();
+    
+// get local date from chrono for database storage purposes
+    let scraped_at = Local::now().date_naive();
+
+// here we find the anchor element and find the value of the href attribute - this should always exist so we can safely unwrap
+    let link = entry.select(&productname_selector).map(|link| {format!("https://amazon.co.uk{}", link.value().attr("href").unwrap())}).collect::<String>();
+
+    vec.push(Product {
+        name,   
+        price: price_text[0].clone(),
+        old_price: Some(price_text[1].clone()),
+        link,                    
+        scraped_at,
+    });
+}
+
+pagenum += 1;
+std::thread::sleep(StdDuration::from_secs(20));
+```
+Note that in the above codeblock we use sleep from the standard library - if we attempt to use `tokio::time::sleep`, the compiler returns an error about holding a non-`Send` future across an await point. 
+
+Now that we've written our code for processing the data we've gathered from the web page, we can wrap what we've written so far in a loop, moving our `Vec<Product>` and `pagenum` declarations to an outer loop that will run infinitely. Next, we'll want to make sure we have somewhere to save our data! We'll want to use a batched transaction here, which thankfully we can do by using `db.begin` and `db.commit`. Check the code out below:
+```rust
+let transaction = db.begin().await.unwrap();
+	
+for product in vec {
+    if let Err(e) = sqlx::query("INSERT INTO 
+        products
+       (name, price, old_price, link, scraped_at)
+       VALUES
+       ($1, $2, $3, $4, $5)
+      ")
+      .bind(product.name)
+      .bind(product.price)
+      .bind(product.old_price)
+      .bind(product.link)
+      .bind(product.scraped_at)
+      .execute(&db)
+      .await
+      .unwrap() {
+        error!("There was an error: {e}");
+        error!("This web scraper will now shut down.");
+        transaction.rollback().await.unwrap();
+        break
+    }
+}
+transaction.commit().await.unwrap();
+```
+All we're doing here is just running a for loop over the list of scraped products and inserting them all into the database, then committing at the end to finalise it.
+
+Now ideally we'll want the scraper to rest for some time so that the pages are given time to update - otherwise, if you comb the pages all the time you will more than likely end up with a huge amount of duplicate data. Let's say we wanted to wanted it to rest until midnight:
+```rust
+use tokio::time::{sleep as TokioSleep, Duration as TokioDuration};
```

Comment:
Same thing. Btw, std Duration should be interchangable with tokio Duration in tokio functions, so maybe no aliases needed on that at all.

Reply:
amended in next commit